cpp/src/stream_compaction/distinct_reduce.cuh

```diff
@@ -0,0 +1,85 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "stream_compaction_common.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/stream_compaction.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <memory>
+
+namespace cudf::detail {
+
+/**
+ * @brief Return the value to initialize reduction results of `spare_reduce_by_row`.
+ *
+ * @param keep A value of `duplicate_keep_option` type, must not be `KEEP_ANY`.
+ * @return The initial reduction value.
+ */
+auto constexpr reduction_init_value(duplicate_keep_option keep)
+{
+  switch (keep) {
+    case duplicate_keep_option::KEEP_FIRST: return std::numeric_limits<size_type>::max();
+    case duplicate_keep_option::KEEP_LAST: return std::numeric_limits<size_type>::min();
+    case duplicate_keep_option::KEEP_NONE: return size_type{0};
+    default: CUDF_UNREACHABLE("KEEP_ANY should not be called with this function");
+  }
+}
+
+/**
+ * @brief Perform a reduction on groups of rows that are compared equal.
+ *
+ * This is essentially a reduce-by-key operation with keys are non-contiguous rows and are compared
+ * equal. Since the keys are sparsely distributed, a hash table is used to find groups of equal
+ * rows.
+ *
+ * Depending on the `keep` parameter, the reduction operation for each row group is:
+ * - If `keep == KEEP_FIRST`: min of row index in the group.
+ * - If `keep == KEEP_LAST`: max of row index in the group.
+ * - If `keep == KEEP_NONE`: group size.
+ *
+ * At the beginning of the operation, the entire output array is filled with a value given by
+ * the `reduction_init_value()` function. Then, the reduction result for each row group is written
+ * into the output array at an index that is the index of an unspecified row in the group.
+ *
+ * @param map The auxiliary map to perform reduction
+ * @param preprocessed_input The preprocessed of the input rows for computing row hashing and row
+ *        comparisons
+ * @param num_rows The number of all input rows
+ * @param has_nulls Indicate whether the input rows has any nulls at any nested levels
+ * @param keep The parameter to determin what type of reduction to perform
+ * @param nulls_equal Flag to specify whether null elements should be considered as equal
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned vector
+ * @return A device_uvector containing the sparse reduction results
+ */
+rmm::device_uvector<size_type> spare_reduce_by_row(
```

Comment:
Why is this is called a "sparse reduction"? I don't really think of a hash map as a traditional "sparse" data structure (like a [sparse matrix](https://en.wikipedia.org/wiki/Sparse_matrix) stored as dictionary of keys or coordinate list). Moreover, we have no knowledge about the number of duplicates in the input data (which could measure the sparsity as the cardinality of the hash map compared to the input size). It may not even be sparse. >50% unique rows would not seem "sparse" to me, but that seems like a very normal input from user data.

I would suggest a name like `row_equivalence_hash_reduction` or `hash_reduce_by_row_equivalence` because that tells you what is being reduced (equivalence classes of rows) and a little bit about the structure used to do the computation (a hash map).

Reply:
`hash_reduce_by_row` sounds good to me.

Reply:
"Reduce by row equivalence" has a specific definition to me that matches this function's behavior, while "reduce by row" does not (it seems to imply the existence of a "reduce by column"). However, this is somewhat subjective and I'll defer to you/others on the final call.

Reply:
A similar example is from thrust, there is `reduce_by_key` that performs reduction on "equivalent" keys :)

Reply:
I'm fine with `hash_reduce_by_row`.